src/file-upload.js

```diff
@@ -49,6 +53,93 @@ export const FilesDragHint = ({isDisplayed}) => html`
   </div>
 `;
 
+/**
+ * @param {Object} data
+ * @param {boolean} isDisplayed
+ * @param {Array<{name: string}>} data.files
+ * @return {lit-html/TemplateResult}
+ */
+export const FilesPanel = ({isDisplayed, files}) => html`
+  <div class="${n('files-panel')}" ?hidden=${!isDisplayed}>
+    ${Toolbar({classNames: [n('files-panel-header')], children: ['Files']})}
+    ${FileList({files})}
+  </div>
+`;
+
+const fileRepeatKey = ({url}) => url;
+
+/**
+ * When `data.files` is empty, renders a "No files" message.
+ * Otherwise renders the file list.
+ * @param {Object} data
+ * @param {Array<{name: string}>} data.files
+ * @return {lit-html/TemplateResult}
+ */
+const FileList = ({files}) =>
+  files.length < 1
+    ? html`
+        <div class="${g('flex-center')} ${n('file-list-empty')}">
+          <div>No files uploaded.</div>
+          ${FileUploadButton()}
+        </div>
+      `
+    : html`
+        <div class="${n('file-list')}">
+          ${repeat(files, fileRepeatKey, FileListItem)}
+        </div>
+      `;
+
+const dispatchInsertFileRef = redispatchAs(g('insert-file-ref'));
+const dispatchDeleteFile = redispatchAs(g('delete-file'));
+
+/**
+ * @param {{name: string}} file
+ * @param {{name: string}} file
+ * @param {number} index
+ * @return {lit-html/TemplateResult}
+ */
+const FileListItem = ({name}, index) => html`
+  <div
+    class="${n('file-list-item')}"
+    data-name="${name}"
+    data-index="${index}"
+    @click="${dispatchInsertFileRef}"
+  >
+    <div class=${n('delete-file-button')} @click=${dispatchDeleteFile}>
+      <span>×</span>
+    </div>
+    <div class="${n('file-list-item-clipped')}">
+      ${name}
+    </div>
+    <div class="${n('file-list-item-unclipped')}">
+      ${name}
+    </div>
+  </div>
+`;
+
+/**
+ * Cascades a click from a parent so it programatically clicks an <input> inside
+ * of it, to propagate a click into a hidden `input[type=file]` element.
+ * @param {Event} e
+ */
+function cascadeInputClick({currentTarget}) {
+  assert(currentTarget.querySelector('input')).click();
+}
+
+const dispatchUploadFiles = redispatchAs(g('upload-files'));
+
+/**
+ * Renders a button to "upload" files--that is, set Blob URLs so they're
+ * accessible from the AMP Ad document.
+ */
+export const FileUploadButton = () => html`
+  <div class="${n('upload-button-container')}" @click="${cascadeInputClick}">
+    <div class="${(n('upload-button'), g('text-button'))}">
```

Comment:
Does this work? shouldn't it be either:

```js
<div class="${`${n('upload-button')} ${g('text-button')}`}">
```

or:

```js
<div class="${[n('upload-button'), g('text-button')].join(' ')}">
```

Reply:
You're right! I didn't notice it because the file upload button doesn't have any special styling. Since it doesn't, would it be better to get rid of `n('upload-button)` altogether or just keep it for labeling's sake?

Reply:
Yeah, we can just remove the classname without styling.